rsocket/RSocketClient.cpp

```diff
@@ -35,28 +35,44 @@ folly::Future<std::unique_ptr<RSocketRequester>> RSocketClient::connect(
   folly::Promise<std::unique_ptr<RSocketRequester>> promise;
   auto future = promise.getFuture();
 
-  connectionFactory_->connect([
-    this,
-    setupParameters = std::move(setupParameters),
-    responder = std::move(responder),
-    keepaliveTimer = std::move(keepaliveTimer),
-    stats = std::move(stats),
-    networkStats = std::move(networkStats),
-    promise = std::move(promise)](
-      std::unique_ptr<DuplexConnection> connection,
-      folly::EventBase& eventBase) mutable {
-    VLOG(3) << "onConnect received DuplexConnection";
-
-    auto rsocket = fromConnection(
-        std::move(connection),
-        eventBase,
-        std::move(setupParameters),
-        std::move(responder),
-        std::move(keepaliveTimer),
-        std::move(stats),
-        std::move(networkStats));
-    promise.setValue(std::move(rsocket));
-  });
+  connectionFactory_->connect()
+      .then([
+        this,
+        setupParameters = std::move(setupParameters),
+        responder = std::move(responder),
+        keepaliveTimer = std::move(keepaliveTimer),
+        stats = std::move(stats),
+        networkStats = std::move(networkStats),
+        promise = std::move(promise)
+      ](ConnectionResult connResult) mutable {
+        VLOG(3) << "onConnect received DuplexConnection";
+        // Ensure we create the RSocketStateMachine in the right evb.
+        connResult.evb.runImmediatelyOrRunInEventBaseThreadAndWait([
```

Comment:
I believe the future will actually resolve on the same thread you call `Promise::setValue()` on, which should be the correct eventbase. So I think this already works, just add a CHECK for being on the current eventbase in the callback?

Otherwise what you can do is in the .then() do
```
return Future::via(evb, [] {
  // .. this will run on evb.
});
```